File: api/product-service/src/main/java/com/technostore/productservice/service/ProductService.java
package com.technostore.productservice.service;

import com.technostore.productservice.dto.ProductDto;
import com.technostore.productservice.dto.SearchProductDto;
import com.technostore.productservice.dto.SortType;
import org.springframework.data.domain.Page;

import javax.servlet.http.HttpServletRequest;
import java.util.List;

public interface ProductService {
    ProductDto getProductById(Long id, HttpServletRequest request);
    Page<SearchProductDto> searchProducts(int numberPage, int sizePage, SortType sort, String word, Integer minRating,
                                          Integer maxRating, Integer minPrice, Integer maxPrice, List<Long> categorie,
                                          Long userId);
}

File: client/network/src/main/java/com/technostore/network/service/ProductService.kt
package com.technostore.network.service

import com.technostore.network.model.product.response.CategoryResponse
import com.technostore.network.model.product.response.ProductDetailResponse
import com.technostore.network.model.product.response.ProductSearchResultResponse
import com.technostore.network.utils.URL
import retrofit2.Response
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface ProductService {
    @GET("product/popular-categories")
    suspend fun getPopularCategories(): Response<List<CategoryResponse>>

    @GET("product/{id}")
    suspend fun getProductById(@Path("id") id: Long): Response<ProductDetailResponse>

    @GET("product/search")
    suspend fun searchProducts(
        @Query("numberPage") numberPage: Int,
        @Query("sizePage") sizePage: Int,
        @Query("word") word: String?,
        @Query("sort") sort: String?,
        @Query("minRating") minRating: Int?,
        @Query("maxRating") maxRating: Int?,
        @Query("categories") categories: String?,
        @Query("minPrice") minPrice: Int?,
        @Query("maxPrice") maxPrice: Int?,
    ): Response<ProductSearchResultResponse>
}